File: mergesort.rb
#! /usr/bin/env ruby

## マージ関数
def merge (arr, s1, e1, s2, e2)
  ret = []
  left = (s1..e1).to_a
  right = (s2..e2).to_a
  ## 与えらた配列の範囲s1~e1, s2~e2が
  ## 昇順にソートされているとして,
  ## それらの部分配列の先頭からピックしながら,
  ## 範囲を線形走査し終えるまで昇順にソートする
  while left.length+right.length > 1
    # 両方の範囲に候補が残っている場合は先頭同士を比較して
    # 小さい方をマージ後に入れる
    if left.length > 0 and right.length > 0
      lv = arr[left[0]]
      rv = arr[right[0]]
      if lv < rv
        ret << arr[left.shift]
      else
        ret << arr[right.shift]
      end
    elsif left.length == 0
      # どちらかが空になっている場合は片方の範囲の先頭を追加する
      ret << arr[right.shift]
    elsif right.length == 0
      ret << arr[left.shift]
    end
  end
  if left.length == 0
    ret << arr[right.shift]
  else
    ret << arr[left.shift]
  end
  i = 0
  for j in s1..e1
    arr[j] = ret[i]
    i += 1
  end
  for j in s2..e2
    arr[j] = ret[i]
    i += 1
  end
end

def msort (arr, left, right)
  w = right-left
  if w > 1
    # 幅が2より大きい場合はそれを更に再帰的に分割してマージする
    center = (left+right)/2-1
    msort arr, left, center
    msort arr, center+1, right
    merge arr, left, center, center+1, right
  elsif w == 1
    # 範囲が幅2になった場合のみソートする
    unless arr[left] < arr[right]
      tmp = arr[left]
      arr[left] = arr[right]
      arr[right] = tmp
    end
  end
end

def sorted? (arr)
  for i in 1..arr.length-1
    return false if arr[i-1] > arr[i]
  end
  return true
end

def main
  arr = (0..100).map { |e| rand 100 }
  p arr
  msort arr, 0, arr.length-1
  p arr
  p "array is sotrted? -> #{sorted? arr}"
end

main

File: mergesort.c
#include <stdio.h>
#include <math.h>
#include <time.h>
#include <stdlib.h>

void merge (int* arr, int s1, int e1, int s2, int e2)
{
  int max_len = e1-s1+1 + e2-s2+1;
  int merged[max_len];
  int i = s1;
  int j = s2;
  int k = 0;
  while (k < max_len) {
    if (i <= e1 && j <= e2) {
      if (arr[i] <= arr[j]) {
        merged[k] = arr[i];
        i++;
      } else if (arr[j] < arr[i]) {
        merged[k] = arr[j];
        j++;
      }
      k++;
    } else if (i > e1) {
      merged[k] = arr[j];
      if (j < e2) {
        merged[++k] = arr[++j];
      }
      j++;
      k++;
    } else if (j > e2) {
      merged[k] = arr[i];
      if (i < e1) {
        merged[++k] = arr[++i];
      }
      i++;
      k++;
    }
  }
  i = s1;
  j = s2;
  k = 0;
  while (i <= e1) {
    arr[i] = merged[k];
    k++;
    i++;
  }
  while (j <= e2) {
    arr[j] = merged[k];
    k++;
    j++;
  }
}

void _mergesort (int *arr, int left, int right)
{
  int w = right-left+1;
  if (w > 2) {
    int center = (int)(left+right)/2;
    _mergesort(arr, left, center);
    _mergesort(arr, center+1, right);
    merge(arr, left, center, center+1, right);
  } else if (w == 2) {
    if (arr[left] > arr[right]) {
      int tmp = arr[left];
      arr[left] = arr[right];
      arr[right] = tmp;
    }
  }
}

void print_array (int* arr, int len)
{
  int i;
  printf("[");
  for (i = 0; i < len; ++i) {
    printf("%i", arr[i]);
    if (i < len-1) {
      printf(",");
    }
  }
  printf("]\n");
}

void random_array (int *addr, int len)
{
  int i = 0;
  srand((unsigned)time(NULL));
  while(i < len) {
    addr[i] = rand()%100;
    i++;
  }
}

int main (int argc, char ** argv)
{
  int len = 100;
  int* arr = (int*)malloc(len*sizeof(int));
  random_array(arr, len);
  print_array(arr, len);
  _mergesort(arr, 0, len-1);
  print_array(arr, len);
}